src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/service/VisitServiceTest.kt

```diff
@@ -0,0 +1,181 @@
+
+  private val visitService: VisitService = VisitService(
+    visitRepository,
+    visitVisitorRepository,
+    offenderBookingRepository,
+    offenderVisitBalanceRepository,
+    offenderVisitBalanceAdjustmentRepository,
+    eventStatusRepository,
+    visitTypeRepository,
+    visitStatusRepository,
+    agencyRepository,
+    telemetryClient,
+  )
+
+  val visitType = VisitType("SCON", "desc")
+  val visitStatus = VisitStatus("SCH", "desc")
+  val eventStatus = EventStatus("SCH", "desc")
+
+  val createVisitRequest = CreateVisitRequest(
+    visitType = "SCON",
+    offenderNo = offenderNo,
+    startTime = LocalDateTime.of(2021, 11, 4, 12, 5),
+    endTime = LocalTime.of(13, 4),
+    prisonId = prisonId,
+    visitorPersonIds = listOf(45L, 46L),
+    decrementBalances = true,
+  )
+
+  @BeforeEach
+  fun setup() {
+    whenever(offenderBookingRepository.findByOffenderNomsIdAndActive(offenderNo, true)).thenReturn(
+      Optional.of(
+        OffenderBooking(
+          bookingId = offenderBookingId,
+          bookingBeginDate = LocalDateTime.now(),
+        )
+      )
+    )
+    whenever(visitTypeRepository.findById(VisitType.pk("SCON"))).thenReturn(Optional.of(visitType))
+    whenever(visitStatusRepository.findById(VisitStatus.pk("SCH"))).thenReturn(Optional.of(visitStatus))
+    whenever(eventStatusRepository.findById(EventStatus.SCHEDULED_APPROVED)).thenReturn(Optional.of(eventStatus))
+    whenever(agencyRepository.findById(prisonId)).thenReturn(Optional.of(AgencyLocation(prisonId, "desc")))
+
+    whenever(offenderVisitBalanceRepository.findById(offenderBookingId)).thenReturn(
+      Optional.of(
+        OffenderVisitBalance(
+          offenderBookingId = offenderBookingId,
+          remainingVisitOrders = 3,
+          remainingPrivilegedVisitOrders = 5,
+        )
+      )
+    )
+  }
+
+  @Nested
+  internal inner class createVisit {
+    @Test
+    fun success() {
+
+      whenever(visitRepository.save(any())).thenReturn(Visit(id = visitId))
+
+      assertThat(visitService.createVisit(createVisitRequest)).isEqualTo(CreateVisitResponse(visitId))
+
+      val balanceArgumentCaptor = ArgumentCaptor.forClass(OffenderVisitBalanceAdjustment::class.java)
+      verify(offenderVisitBalanceAdjustmentRepository).save(balanceArgumentCaptor.capture())
+      val balanceArgument = balanceArgumentCaptor.value
+      assertThat(balanceArgument.adjustReasonCode).isEqualTo(VisitOrderAdjustmentReason.VISIT_ORDER_ISSUE)
+      assertThat(balanceArgument.remainingVisitOrders).isEqualTo(-1)
+      assertThat(balanceArgument.remainingPrivilegedVisitOrders).isNull()
+      assertThat(balanceArgument.commentText).isEqualTo("Created by PVB3 for an on-line visit booking")
+
+      val visitArgumentCaptor = ArgumentCaptor.forClass(Visit::class.java)
+      verify(visitRepository).save(visitArgumentCaptor.capture())
+      val visit = visitArgumentCaptor.value
+      assertThat(visit.visitDate).isEqualTo(LocalDate.of(2021, 11, 4))
```

Comment:
LocalDate.parse("") tends to be more readable in tests the using `of`

Strings are more readable then numbers on the correct order